Implement "join pane into next"

tcard opened the pull request:
Release Notes:

- Added "join pane into next" action ([#12409](https://github.com/zed-industries/zed/issues/12409))



https://github.com/zed-industries/zed/assets/727422/00cc8599-e5d6-4fb8-9f0d-9b232d2210ef


---

This is just the basic implementation. I wanted to get some feedback/validation before going forwards.

I think the next steps may be:

* Add test. Probably by adding something to the existing `test_panels` (sic).
* Deduplicate pane items when joining. VSCode does this. Without this, since items are cloned on pane split, it's then annoying to join panes and get both the original item and its previously cloned item into the same pane.
  * The `Item` trait could get a `should_deduplicate_with(&dyn Pane)` or similar. Each `Item` would decide if it's "the same" as the other item following its own logic, e. g. `Editor` would deduplicate if it has the same buffers as the other `Item`.

tcard committed:
Implement "join pane into next"

tcard commented on crates/workspace/src/workspace.rs:
```diff
@@ -2655,6 +2656,20 @@ impl Workspace {
+        let Some(next_pane) = next_pane else {
+            return;
+        };
+        let item_ids: Vec<EntityId> = pane.read(cx).items().map(|item| item.item_id()).collect();
```
I did this because just iterating on `items()` would throw errors like:

```
error[E0502]: cannot borrow `*cx` as mutable because it is also borrowed as immutable
    --> crates/workspace/src/workspace.rs:2668:13
     |
2666 |         let pane_read_handle = pane.read(cx);
     |                                          -- immutable borrow occurs here
2667 |         for item in pane_read_handle.items() {
     |                     ------------------------ immutable borrow later used here
2668 |             self.move_item(pane.clone(), next_pane.clone(), item.item_id(), 0, cx);
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```

I'm not extremely comfortable with Rust, so I welcome suggestions to remove this allocation if possible.

SomeoneToIgnore replied:
I'm afraid there's no better way for this particular part of the code, but this is not a hot path, so why not.

tcard commented:
@cla-bot check

canadaduane commented:
A couple of questions:
- would it make sense to have a "move all" to next pane?
- what happens if the same buffer is open in two split panes?
- would it be better to add a direction, similar to "pane: split left", "pane: split right", "pane: split down" etc? If so, it would be a nice inverse operation to each of the split actions. You could "pane: split left", then "pane: join right" to have it merge back where it came from.

tcard commented:
@canadaduane 
 
>     * would it make sense to have a "move all" to next pane?

This would be the same as the "join into next" but without closing the pane? Empty panes like this aren't present at the moment (as far as I can tell). I don't see much need to introduce them for this.

>     * what happens if the same buffer is open in two split panes?

You get two (or more) items (tabs) with the same buffer in the next pane.

VSCode prevents this by removing one of the tabs. I refer to this in the pull request description.

>     * would it be better to add a direction, similar to "pane: split left", "pane: split right", "pane: split down" etc, instead of the slightly ambiguous "next" direction? If so, it would be a nice inverse operation to each of the split actions. You could "pane: split left", then "pane: join right" to have it merge back where it came from.

Indeed. I just reproduced (minus the deduplication I mentioned) VSCode's behavior, but I agree this would flexibility would be nice. I could give it a go, although the currently implemented behavior already scratches my particular itch .

> @tcard : in addition to doing this by opening the menu as happens now, for me it would make sense to give the possibility if we find ourselves with more bread open a button that allows us to do the same thing.

@Angelk90 I think having buttons for all possible actions may clutter the UI too much. Maybe it could be an option in this menu though?

<img width="235" alt="Screenshot 2024-06-02 at 17 14 35" src="https://github.com/zed-industries/zed/assets/727422/6127e4df-95f2-4c5a-a25d-24c975df7eba">

canadaduane commented:
>>    * would it make sense to have a "move all" to next pane?

> This would be the same as the "join into next" but without closing the pane? Empty panes like this aren't present at the moment (as far as I can tell). I don't see much need to introduce them for this.

@tcard I think I did a poor job of explaining myself here. I may be using incorrect terminology. I noted in the video that each file tab (pane?) that you demonstrate moves one-at-a-time to the next split area ("join into next"). I was wondering if there is a command that could move all files (panes?) to the next split area (or possibly directionally). Basically a way to close/remove an entire split area, but describe where to send all of its panes.

SomeoneToIgnore commented on crates/workspace/src/workspace.rs:
```diff
@@ -2655,6 +2656,20 @@ impl Workspace {
+    pub fn join_pane_into_next(&mut self, pane: View<Pane>, cx: &mut ViewContext<Self>) {
+        let next_pane = self
+            .find_pane_in_direction(SplitDirection::Right, cx)
+            .or_else(|| self.find_pane_in_direction(SplitDirection::Left, cx));
```
Could we actually look for all split directions here?
I found this feature useful for closing void panes (to get one, drag and drop some unsupported file, e.g. *.ttf onto Zed)

![image](https://github.com/zed-industries/zed/assets/2690773/8915da3d-6b4c-470c-ba75-409cd90cb8ab)

but the current code cannot close certain combinations of these, e.g.

![image](https://github.com/zed-industries/zed/assets/2690773/13ff1f9f-b9cd-4f88-8bbb-8ae96658c30b)

SomeoneToIgnore reviewed (commented):
Sorry for the delayed review, but not I'm responsible for this so we have a chance to move forward 🙂 

Should we add a test for this, similar to 

https://github.com/zed-industries/zed/blob/c0ac9e752b82935a38c809e4d8a87fb14c623921/crates/workspace/src/workspace.rs#L5537

multiple panes can be created and a new action called.

SomeoneToIgnore commented:
I'll close this as stale for now, but feel free to ping me in a new PR as it's reviewed already.

tcard commented:
I've been a bit too busy lately but I intend to come back to this soon.
